pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -139,58 +145,88 @@ public RebalanceResult rebalance(TableConfig tableConfig, RebalanceSegmentStrate
         return result;
       }
 
       // if ideal state needs to change, get the next 'safe' state (based on whether downtime is OK or not)
       IdealState nextIdealState = getNextState(currentIdealState, targetIdealState, rebalanceConfig);
+      LOGGER.debug("Ideal state after making changes to partition map of all segments");
+      prettyPrintIdealState(nextIdealState);
 
       // If the ideal state is large enough, enable compression
       if (HelixHelper.MAX_PARTITION_COUNT_IN_UNCOMPRESSED_IDEAL_STATE < nextIdealState.getPartitionSet().size()) {
         nextIdealState.getRecord().setBooleanField("enableCompression", true);
       }
 
-      // Check version and set ideal state
+      // Check version and set ideal state to nextIdealState
       try {
-        LOGGER.info("Updating IdealState for table {}", tableNameWithType);
+        LOGGER.info("Going to update current IdealState in ZK for table {}, current version {} and creation time {}",
+            tableNameWithType, currentIdealState.getRecord().getVersion(), currentIdealState.getRecord().getCreationTime());
         if (zkBaseDataAccessor
             .set(idealStateKey.getPath(), nextIdealState.getRecord(), currentIdealState.getRecord().getVersion(),
                 AccessOption.PERSISTENT)) {
+          LOGGER.debug("Successfully persisted the ideal state for table {} in ZK. Will wait for External view to converge",
+              tableNameWithType);
+          ++_rebalancerStats.updatestoIdealStateInZK;
           // if we succeeded, wait for the change to stabilize
           waitForStable(tableNameWithType);
           // clear retries as it tracks failures with each idealstate update attempt
           retries = 0;
+          LOGGER.debug("External view converged for the change in ideal state for table {}. Will start the next iteration now",
+              tableNameWithType);
           continue;
         }
-        // in case of any error, we retry a bounded number of types
+        // in case of any error, we retry a bounded number of times
       } catch (ZkBadVersionException e) {
+        // we will go back in the loop and reattempt by recomputing the target ideal state
         LOGGER.warn("Version changed while updating ideal state for resource: {}", tableNameWithType);
       } catch (Exception e) {
-        LOGGER.warn("Caught exception while updating ideal state for resource: {}", tableNameWithType, e);
+        if (e instanceof IllegalStateException && e.getCause() instanceof ExternalViewErrored) {
+          LOGGER.error("External view reported error for table {} after updating ideal state", tableNameWithType);
+          // remove the cause as it is private and is only used to detect exact error within this class
+          throw new IllegalStateException(e.getMessage());
+        } else {
+          LOGGER.warn("Caught exception while or after updating ideal state for resource: {}", tableNameWithType, e);
```

Comment:
It makes total sense to continue (restart by regetting the current ideal state from ZK and recomputing the target) if the atomic operation failed in ZK due to version mismatch. 

However, if there is any other unpredictable exception (other than Illegal state exception due to error state in external view), I suggest we still bail out

Reply:
agreed

Reply:
Done